File: src/15.cpp
class Solution {
public:
    void twoSum(int left, int right, int target, vector<int> &numbers, vector<vector<int>> &result)
    {
        while (left < right) {
            if (numbers[left] + numbers[right] == target) {
                vector<int> Three;
                Three.push_back(-target);
                Three.push_back(numbers[left]);
                Three.push_back(numbers[right]);
                result.push_back(Three);
                
                left++;
                right--;
                
                while (left < right && numbers[left] == numbers[left - 1]) {
                    left++;
                }
                while (left < right && numbers[right] == numbers[right + 1]) {
                    right--;
                }
                
            } else if (numbers[left] + numbers[right] < target) {
                left++;
            } else {
                right--;
            }
        }
    }
     
    vector<vector<int>> threeSum(vector<int> &numbers) {
        // write your code here
        vector<vector<int>> result;
        if (numbers.size() < 3) {
            return result;
        }
        
        sort(numbers.begin(), numbers.end());

        for (int i = 0; i < numbers.size(); i++) {
            if (i > 0 && numbers[i] == numbers[i - 1]) {
                continue;
            }
            int target = -numbers[i];
            int left = i + 1;
            int right = numbers.size() - 1;
            twoSum(left, right, target, numbers, result);
        }
        return result;
    }
};

File: src/15.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ##corner case
        res = []
        if len(nums) < 3:
            return res
        
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            a = i + 1
            b = len(nums) - 1
            while a < b:
                if nums[a] + nums[b] == -nums[i]:
                    temp = []
                    temp.append(nums[i])
                    temp.append(nums[a])
                    temp.append(nums[b])
                    res.append(temp)
                    a += 1
                    b -= 1
                    while a < b and nums[a] == nums[a - 1]:
                        a += 1
                    while a < b and nums[b] == nums[b + 1]:
                        b -= 1
                elif nums[a] + nums[b] > -nums[i]:
                    b -= 1
                else:
                    a += 1
        
        return res